File: opine/backend/scripts/updateResponsesFromExcel.js
const mongoose = require('mongoose');
const XLSX = require('xlsx');
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '../.env') });

const SurveyResponse = require('../models/SurveyResponse');
const User = require('../models/User');

const EXCEL_FILE_PATH = '/var/www/reports/Vijay_CAPI & CATI_Server IDs for valid and Reject_5th Feb (1).xlsx';
const REVIEWER_EMAIL = 'ajayadarsh@gmail.com';
const REJECTION_REASON = 'Manual Rejection on 6 feb 2026';

async function updateResponsesFromExcel() {
  try {
    // Connect to MongoDB
    await mongoose.connect(process.env.MONGODB_URI);
    console.log('✅ Connected to MongoDB');

    // Find the reviewer user
    const reviewer = await User.findOne({ email: REVIEWER_EMAIL });
    if (!reviewer) {
      throw new Error(`Reviewer not found: ${REVIEWER_EMAIL}`);
    }
    console.log(`✅ Found reviewer: ${REVIEWER_EMAIL} (ID: ${reviewer._id})`);

    // Load Excel file
    const workbook = XLSX.readFile(EXCEL_FILE_PATH);
    console.log('✅ Loaded Excel file');
    console.log(`   Sheets: ${workbook.SheetNames.join(', ')}`);

    const sheets = ['CATI', 'CAPI'];
    let totalProcessed = 0;
    let totalApproved = 0;
    let totalRejected = 0;
    let totalNotFound = 0;
    let totalErrors = 0;
    const notFoundIds = [];
    const errorIds = [];

    for (const sheetName of sheets) {
      console.log(`\n📊 Processing sheet: ${sheetName}`);
      
      if (!workbook.Sheets[sheetName]) {
        console.log(`⚠️  Sheet ${sheetName} not found, skipping...`);
        continue;
      }

      // Convert sheet to JSON
      const worksheet = workbook.Sheets[sheetName];
      const data = XLSX.utils.sheet_to_json(worksheet, { raw: false });
      
      if (!data || data.length === 0) {
        console.log(`⚠️  Sheet ${sheetName} is empty, skipping...`);
        continue;
      }

      // Find column names (handle case variations)
      let responseIdCol = null;
      let statusCol = null;
      
      const firstRow = data[0];
      for (const key in firstRow) {
        const keyLower = key.toLowerCase().trim();
        if (keyLower.includes('response') && keyLower.includes('id')) {
          responseIdCol = key;
        }
        if (keyLower.includes('revised') && keyLower.includes('status')) {
          statusCol = key;
        }
      }

      if (!responseIdCol || !statusCol) {
        console.log(`⚠️  Could not find required columns in sheet ${sheetName}, skipping...`);
        console.log(`   Available columns: ${Object.keys(firstRow).join(', ')}`);
        continue;
      }

      console.log(`   Found columns: Response ID = "${responseIdCol}", Status = "${statusCol}"`);
      console.log(`   Total rows: ${data.length}`);

      // Process each row
      let sheetProcessed = 0;
      let sheetApproved = 0;
      let sheetRejected = 0;
      let sheetNotFound = 0;
      let sheetErrors = 0;

      for (const row of data) {
        const responseId = row[responseIdCol]?.toString().trim();
        const status = row[statusCol]?.toString().trim();

        if (!responseId || !status) {
          continue; // Skip empty rows
        }

        // Normalize status
        const normalizedStatus = status.toLowerCase();
        if (normalizedStatus !== 'approved' && normalizedStatus !== 'rejected') {
          console.log(`⚠️  Invalid status "${status}" for response ${responseId}, skipping...`);
          continue;
        }

        try {
          // Find the response
          const response = await SurveyResponse.findOne({ responseId: responseId });

          if (!response) {
            console.log(`❌ Response not found: ${responseId}`);
            notFoundIds.push(responseId);
            sheetNotFound++;
            totalNotFound++;
            continue;
          }

          // ⚠️ CRITICAL: Only update status from Excel, preserve existing reviewer if it exists
          // If response already has a reviewer, keep it - don't overwrite
          // Only update responses that are in the Excel file
          const existingReviewer = response.verificationData?.reviewer;
          const hasExistingReviewer = existingReviewer && existingReviewer.toString();

          // Prepare update data - only update status
          const updateData = {
            $set: {
              status: normalizedStatus === 'approved' ? 'Approved' : 'Rejected',
              updatedAt: new Date()
            },
            $unset: {
              reviewAssignment: ''
            }
          };

          // Only set reviewer if it doesn't already exist
          if (!hasExistingReviewer) {
            updateData.$set['verificationData.reviewer'] = reviewer._id;
            updateData.$set['verificationData.reviewedAt'] = new Date();
          }
          // If reviewer exists, preserve it and only update reviewedAt if needed
          else {
            // Preserve existing reviewer - don't change it
            // Only update reviewedAt if it doesn't exist
            if (!response.verificationData?.reviewedAt) {
              updateData.$set['verificationData.reviewedAt'] = new Date();
            }
          }

          // Add rejection reason if rejected (only if not already set)
          if (normalizedStatus === 'rejected') {
            if (!response.verificationData?.feedback || 
                response.verificationData.feedback === REJECTION_REASON) {
              updateData.$set['verificationData.feedback'] = REJECTION_REASON;
            }
            // If feedback exists and is different, preserve it
          } else {
            // For approved, only set empty feedback if none exists
            if (!response.verificationData?.feedback) {
              updateData.$set['verificationData.feedback'] = '';
            }
          }

          // Preserve existing verification criteria if they exist
          if (response.verificationData?.criteria) {
            updateData.$set['verificationData.criteria'] = response.verificationData.criteria;
          }
          
          // Preserve all other existing verificationData fields
          if (response.verificationData) {
            // Preserve reviewHistory if it exists
            if (response.verificationData.reviewHistory) {
              updateData.$set['verificationData.reviewHistory'] = response.verificationData.reviewHistory;
            }
            // Preserve originalReviewer if it exists
            if (response.verificationData.originalReviewer) {
              updateData.$set['verificationData.originalReviewer'] = response.verificationData.originalReviewer;
            }
            // Preserve previousReviewer if it exists
            if (response.verificationData.previousReviewer) {
              updateData.$set['verificationData.previousReviewer'] = response.verificationData.previousReviewer;
            }
          }

          // Update the response
          const updatedResponse = await SurveyResponse.findOneAndUpdate(
            { _id: response._id },
            updateData,
            { new: true, runValidators: false }
          );

          if (updatedResponse) {
            if (normalizedStatus === 'approved') {
              sheetApproved++;
              totalApproved++;
            } else {
              sheetRejected++;
              totalRejected++;
            }
            sheetProcessed++;
            totalProcessed++;

            if (sheetProcessed % 100 === 0) {
              console.log(`   Processed ${sheetProcessed} responses...`);
            }
          } else {
            console.log(`❌ Failed to update response: ${responseId}`);
            errorIds.push(responseId);
            sheetErrors++;
            totalErrors++;
          }
        } catch (error) {
          console.error(`❌ Error processing response ${responseId}:`, error.message);
          errorIds.push(responseId);
          sheetErrors++;
          totalErrors++;
        }
      }

      console.log(`\n✅ Sheet ${sheetName} completed:`);
      console.log(`   Total processed: ${sheetProcessed}`);
      console.log(`   Approved: ${sheetApproved}`);
      console.log(`   Rejected: ${sheetRejected}`);
      console.log(`   Not found: ${sheetNotFound}`);
      console.log(`   Errors: ${sheetErrors}`);
    }

    console.log(`\n📊 ===== FINAL SUMMARY =====`);
    console.log(`Total processed: ${totalProcessed}`);
    console.log(`Total approved: ${totalApproved}`);
    console.log(`Total rejected: ${totalRejected}`);
    console.log(`Total not found: ${totalNotFound}`);
    console.log(`Total errors: ${totalErrors}`);

    if (notFoundIds.length > 0) {
      console.log(`\n⚠️  Not found response IDs (first 20):`);
      console.log(notFoundIds.slice(0, 20).join(', '));
      if (notFoundIds.length > 20) {
        console.log(`   ... and ${notFoundIds.length - 20} more`);
      }
    }

    if (errorIds.length > 0) {
      console.log(`\n⚠️  Error response IDs (first 20):`);
      console.log(errorIds.slice(0, 20).join(', '));
      if (errorIds.length > 20) {
        console.log(`   ... and ${errorIds.length - 20} more`);
      }
    }

    await mongoose.connection.close();
    console.log('\n✅ Script completed successfully');
    process.exit(0);
  } catch (error) {
    console.error('❌ Fatal error:', error);
    await mongoose.connection.close();
    process.exit(1);
  }
}

// Run the script
updateResponsesFromExcel();

File: opine/backend/scripts/updateResponsesFromExcel.py
#!/usr/bin/env python3
"""
Script to update survey responses from Excel file
Processes both CATI and CAPI sheets
"""
import os
import sys
from datetime import datetime
import pandas as pd
from pymongo import MongoClient
from dotenv import load_dotenv

# Load environment variables
load_dotenv(os.path.join(os.path.dirname(__file__), '../.env'))

# Configuration
EXCEL_FILE_PATH = '/var/www/reports/Vijay_CAPI & CATI_Server IDs for valid and Reject_5th Feb (1).xlsx'
REVIEWER_EMAIL = 'ajayadarsh@gmail.com'
REJECTION_REASON = 'Manual Rejection on 6 feb 2026'

def main():
    # Connect to MongoDB
    mongodb_uri = os.getenv('MONGODB_URI')
    if not mongodb_uri:
        print('❌ MONGODB_URI not found in environment variables')
        sys.exit(1)
    
    client = MongoClient(mongodb_uri)
    db = client.get_database()
    survey_responses = db.surveyresponses
    users = db.users
    
    print('✅ Connected to MongoDB')
    
    # Find the reviewer user
    reviewer = users.find_one({'email': REVIEWER_EMAIL})
    if not reviewer:
        print(f'❌ Reviewer not found: {REVIEWER_EMAIL}')
        sys.exit(1)
    
    reviewer_id = reviewer['_id']
    print(f'✅ Found reviewer: {REVIEWER_EMAIL} (ID: {reviewer_id})')
    
    # Load Excel file
    try:
        xls = pd.ExcelFile(EXCEL_FILE_PATH)
        print('✅ Loaded Excel file')
        print(f'   Sheets: {xls.sheet_names}')
    except Exception as e:
        print(f'❌ Error loading Excel file: {e}')
        sys.exit(1)
    
    sheets = ['CATI', 'CAPI']
    total_stats = {
        'processed': 0,
        'approved': 0,
        'rejected': 0,
        'not_found': 0,
        'errors': 0
    }
    not_found_ids = []
    error_ids = []
    
    for sheet_name in sheets:
        print(f'\n📊 Processing sheet: {sheet_name}')
        
        try:
            df = pd.read_excel(xls, sheet_name=sheet_name)
        except Exception as e:
            print(f'⚠️  Error reading sheet {sheet_name}: {e}')
            continue
        
        # Find column names (handle case variations)
        response_id_col = None
        status_col = None
        
        for col in df.columns:
            col_lower = str(col).lower().strip()
            if 'response' in col_lower and 'id' in col_lower:
                response_id_col = col
            if 'revised' in col_lower and 'status' in col_lower:
                status_col = col
        
        if not response_id_col or not status_col:
            print(f'⚠️  Could not find required columns in sheet {sheet_name}')
            print(f'   Available columns: {list(df.columns)}')
            continue
        
        print(f'   Found columns: Response ID = "{response_id_col}", Status = "{status_col}"')
        print(f'   Total rows: {len(df)}')
        
        sheet_stats = {
            'processed': 0,
            'approved': 0,
            'rejected': 0,
            'not_found': 0,
            'errors': 0
        }
        
        # Process each row
        for idx, row in df.iterrows():
            response_id = str(row[response_id_col]).strip() if pd.notna(row[response_id_col]) else None
            status = str(row[status_col]).strip() if pd.notna(row[status_col]) else None
            
            if not response_id or not status or response_id == 'nan' or status == 'nan':
                continue  # Skip empty rows
            
            # Normalize status
            status_lower = status.lower()
            if status_lower not in ['approved', 'rejected']:
                print(f'⚠️  Invalid status "{status}" for response {response_id}, skipping...')
                continue
            
            try:
                # Find the response
                response = survey_responses.find_one({'responseId': response_id})
                
                if not response:
                    print(f'❌ Response not found: {response_id}')
                    not_found_ids.append(response_id)
                    sheet_stats['not_found'] += 1
                    total_stats['not_found'] += 1
                    continue
                
                # Prepare update data
                new_status = 'Approved' if status_lower == 'approved' else 'Rejected'
                
                update_data = {
                    '$set': {
                        'status': new_status,
                        'verificationData.reviewer': reviewer_id,
                        'verificationData.reviewedAt': datetime.utcnow(),
                        'updatedAt': datetime.utcnow()
                    },
                    '$unset': {
                        'reviewAssignment': ''
                    }
                }
                
                # Add rejection reason if rejected
                if status_lower == 'rejected':
                    update_data['$set']['verificationData.feedback'] = REJECTION_REASON
                else:
                    # For approved, preserve existing feedback or set empty
                    if 'verificationData' in response and 'feedback' in response.get('verificationData', {}):
                        # Keep existing feedback
                        pass
                    else:
                        update_data['$set']['verificationData.feedback'] = ''
                
                # Preserve existing verification criteria if they exist
                if 'verificationData' in response and 'criteria' in response.get('verificationData', {}):
                    update_data['$set']['verificationData.criteria'] = response['verificationData']['criteria']
                
                # Update the response
                result = survey_responses.update_one(
                    {'_id': response['_id']},
                    update_data
                )
                
                if result.modified_count > 0 or result.matched_count > 0:
                    if status_lower == 'approved':
                        sheet_stats['approved'] += 1
                        total_stats['approved'] += 1
                    else:
                        sheet_stats['rejected'] += 1
                        total_stats['rejected'] += 1
                    
                    sheet_stats['processed'] += 1
                    total_stats['processed'] += 1
                    
                    if sheet_stats['processed'] % 100 == 0:
                        print(f'   Processed {sheet_stats["processed"]} responses...')
                else:
                    print(f'❌ Failed to update response: {response_id}')
                    error_ids.append(response_id)
                    sheet_stats['errors'] += 1
                    total_stats['errors'] += 1
                    
            except Exception as error:
                print(f'❌ Error processing response {response_id}: {str(error)}')
                error_ids.append(response_id)
                sheet_stats['errors'] += 1
                total_stats['errors'] += 1
        
        print(f'\n✅ Sheet {sheet_name} completed:')
        print(f'   Total processed: {sheet_stats["processed"]}')
        print(f'   Approved: {sheet_stats["approved"]}')
        print(f'   Rejected: {sheet_stats["rejected"]}')
        print(f'   Not found: {sheet_stats["not_found"]}')
        print(f'   Errors: {sheet_stats["errors"]}')
    
    print(f'\n📊 ===== FINAL SUMMARY =====')
    print(f'Total processed: {total_stats["processed"]}')
    print(f'Total approved: {total_stats["approved"]}')
    print(f'Total rejected: {total_stats["rejected"]}')
    print(f'Total not found: {total_stats["not_found"]}')
    print(f'Total errors: {total_stats["errors"]}')
    
    if not_found_ids:
        print(f'\n⚠️  Not found response IDs (first 20):')
        print(', '.join(not_found_ids[:20]))
        if len(not_found_ids) > 20:
            print(f'   ... and {len(not_found_ids) - 20} more')
    
    if error_ids:
        print(f'\n⚠️  Error response IDs (first 20):')
        print(', '.join(error_ids[:20]))
        if len(error_ids) > 20:
            print(f'   ... and {len(error_ids) - 20} more')
    
    client.close()
    print('\n✅ Script completed successfully')

if __name__ == '__main__':
    main()






